tracy-client/src/gpu.rs

```diff
@@ -0,0 +1,201 @@
+    pub fn new_gpu_context(
+        self,
+        name: Option<&str>,
+        ty: GpuContextType,
+        gpu_timestamp: i64,
+        period: f32,
+    ) -> Option<GpuContext> {
+        #[cfg(feature = "enable")]
+        {
+            // We use a mutex to lock the context index to prevent races when using fetch_add.
+            let mut context_index_guard = GPU_CONTEXT_INDEX.lock().unwrap();
+            if *context_index_guard == 255 {
+                return None;
+            }
+            let context = *context_index_guard;
+            *context_index_guard += 1;
+            drop(context_index_guard);
+
+            unsafe {
+                sys::___tracy_emit_gpu_new_context(sys::___tracy_gpu_new_context_data {
+                    gpuTime: gpu_timestamp,
+                    period,
+                    context,
+                    flags: 0,
+                    type_: ty as u8,
+                })
+            };
+
+            if let Some(name) = name {
+                unsafe {
+                    sys::___tracy_emit_gpu_context_name_serial(
+                        sys::___tracy_gpu_context_name_data {
+                            context,
+                            name: name.as_ptr().cast(),
+                            len: name.len().try_into().unwrap_or(u16::MAX),
+                        },
+                    )
+                }
+            }
+
+            Some(GpuContext {
+                client: self,
+                context,
+                span_index: Arc::new(AtomicU16::new(0)),
+            })
+        }
+        #[cfg(not(feature = "enable"))]
+        Some(GpuContext {})
+    }
+}
+
+impl GpuContext {
+    ///
+    pub fn span_alloc(&self, name: &str, function: &str, file: &str, line: u32) -> GpuSpan {
+        #[cfg(feature = "enable")]
+        {
+            let srcloc = unsafe {
+                sys::___tracy_alloc_srcloc_name(
+                    line,
+                    file.as_ptr().cast(),
+                    file.len(),
+                    function.as_ptr().cast(),
+                    function.len(),
+                    name.as_ptr().cast(),
+                    name.len(),
+                )
+            };
+
+            // Allocate two ids, one for the start value, one for the end.
+            let query_id = self.span_index.fetch_add(2, Ordering::Relaxed);
+
+            unsafe {
+                sys::___tracy_emit_gpu_zone_begin_alloc_serial(sys::___tracy_gpu_zone_begin_data {
+                    srcloc,
+                    queryId: query_id,
+                    context: self.context,
+                })
+            };
+
+            GpuSpan {
+                _client: self.client.clone(),
+                context: self.context,
+                query_id,
+                ended: false,
+            }
+        }
+        #[cfg(not(feature = "enable"))]
+        GpuSpan {}
+    }
+}
+
+impl GpuSpan {
+    ///
+    pub fn end_zone(&mut self) -> Option<()> {
```

Comment:
What's the significance of this call is if we're passing timestamps explicitly anyway? Is tracy overloading this API to track the CPU time involved in recording commands as well?

(should probably be answered in the doc comment, not just the PR)

Reply:
Tracy correlates the gpu side timestamps with cpu side timestamps - each creation call is where tracy pulls the cpu side timestamps from.